wgpu-core/src/device/queue.rs

```diff
@@ -143,12 +143,56 @@ crate::impl_storage_item!(Queue);
 impl Drop for Queue {
     fn drop(&mut self) {
         resource_log!("Drop {}", self.error_ident());
+
+        let last_successful_submission_index = self
+            .device
+            .last_successful_submission_index
+            .load(Ordering::Acquire);
+
+        let fence = self.device.fence.read();
+        let wait_res = unsafe {
+            self.device.raw().wait(
+                fence.as_ref(),
+                last_successful_submission_index,
+                crate::device::CLEANUP_WAIT_MS,
+            )
+        };
+        drop(fence);
+
+        match wait_res {
+            Ok(true) => {}
+            // Note: If we don't panic here we are in UB land (destroying resources while they are still in use by the GPU).
+            Ok(false) => {
+                panic!("We timed out while waiting on the last successful submission to complete!");
+            }
+            Err(e) => {
+                panic!(
+                    "We ran into an error while waiting on the last successful submission to complete! - {e}"
+                );
+            }
+        }
+
+        let snatch_guard = self.device.snatchable_lock.read();
+        let (submission_closures, mapping_closures, queue_empty) =
+            self.maintain(last_successful_submission_index, &snatch_guard);
+        drop(snatch_guard);
```

Comment:
nitpick: Manual `drop`s of locks are sometimes necessary, but this doesn't seem like a case where it _is_ necessary.	

suggestion: I think it's strictly easier to perform assignment of these variables inside a block, so we can't forget to unlock locks.

```suggestion
        let wait_res;
        {
            let fence = self.device.fence.read();
            wait_res = unsafe {
                self.device.raw().wait(
                    fence.as_ref(),
                    last_successful_submission_index,
                    crate::device::CLEANUP_WAIT_MS,
                )
            };
        }

        match wait_res {
            Ok(true) => {}
            // Note: If we don't panic here we are in UB land (destroying resources while they are still in use by the GPU).
            Ok(false) => {
                panic!("We timed out while waiting on the last successful submission to complete!");
            }
            Err(e) => {
                panic!(
                    "We ran into an error while waiting on the last successful submission to complete! - {e}"
                );
            }
        }

        let (submission_closures, mapping_closures, queue_empty);
        {
            let snatch_guard = self.device.snatchable_lock.read();
            (submission_closures, mapping_closures, queue_empty) =
                self.maintain(last_successful_submission_index, &snatch_guard);
        }
```

Reply:
I prefer to have the explicit `drop` calls rather than adding new scopes to trigger `drop` implicitly.